driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -181,90 +189,219 @@ public boolean isClosed() {
+                return myResponse;
+            }
+
+            try {
+                localLatch.await();
+            } catch (InterruptedException e) {
+                throw new MongoInterruptedException("Interrupted while reading from stream", e);
+            }
+
+            localLatch = readingPhase.get();
         }
     }
 
     @Override
     public void sendMessageAsync(final List<ByteBuf> byteBuffers, final int lastRequestId, final SingleResultCallback<Void> callback) {
-        notNull("open", stream, callback);
+        notNull("stream is open", stream, callback);
+
+        if (isClosed()) {
+            callback.onResult(null, new MongoSocketClosedException("Can not read from a closed socket", getServerAddress()));
+        }
+
         if (LOGGER.isTraceEnabled()) {
-            LOGGER.trace(format("Queuing send message: %s. ([%s] %s)", lastRequestId, getId(), serverId));
+            LOGGER.trace(format("Queuing send message: %s. ([%s])", lastRequestId, getId()));
+        }
+
+        SendMessageRequest sendMessageRequest = new SendMessageRequest(byteBuffers, lastRequestId, errorHandlingCallback(callback, LOGGER));
+
+        boolean mustWrite = false;
+        writerLock.lock();
+        try {
+            if (isWriting) {
+                writeQueue.add(sendMessageRequest);
+            } else {
+                isWriting = true;
+                mustWrite = true;
+            }
+
+        } finally {
+            writerLock.unlock();
+        }
+
+        if (mustWrite) {
+            writeAsync(sendMessageRequest);
         }
-        writeQueue.add(new SendMessageAsync(byteBuffers, lastRequestId, errorHandlingCallback(callback, LOGGER)));
-        processPendingWrites();
+    }
+
+    private void writeAsync(final SendMessageRequest request) {
+        stream.writeAsync(request.getByteBuffers(), new AsyncCompletionHandler<Void>() {
+            @Override
+            public void completed(final Void v) {
+                SendMessageRequest nextMessage = null;
+                writerLock.lock();
+                try {
+                    nextMessage = writeQueue.poll();
+                    if (nextMessage == null) {
+                        isWriting = false;
+                    }
+                } finally {
+                    writerLock.unlock();
+                }
+
+                connectionListener.messagesSent(new ConnectionMessagesSentEvent(getId(), request.getMessageId(),
+                                                                                getTotalRemaining(request.getByteBuffers())));
+                request.getCallback().onResult(null, null);
+
+                if (nextMessage != null) {
+                    writeAsync(nextMessage);
+                }
+            }
+
+            @Override
+            public void failed(final Throwable t) {
+                close();
```

Comment:
Do we need to do anything with the isWriting boolean here? I assume no because the connection has failed...